### PR TITLE
Print warning about kubelet not running on darwin via local-up-cluster.sh

### PR DESCRIPTION
**Pull request by alejandroEsc**
**What this PR does / why we need it**: As of the time of opening this PR, near the release of v.1.5, using `./hack/local-up-cluster.sh` claims to succeed and run kubelet, but logs, processes, and attempts to hit the service demonstrate that kubelet is NOT running. The issue has been found to be the hard requirements of cadvisor forcing kubelet to force exit when it is found to be unsupported as in the case of DARWIN.  

My goal here is to bring the issue to attention to others and also help other devs avoid the issue.

**Release note**:
```NONE```

kubelet.go and kubelet's server.go file has minimal logic that keeps kubelet from breaking just because osx has not cadvisor support.

### CODE REVIEW

**Comment by alejandroEsc** [comment]
@ixdy feel free to weigh in here as well.

**Comment by alejandroEsc** [comment]
@k8s-bot gci gce e2e test this

**Comment by alejandroEsc** [comment]
@kubernetes/api-reviewers have waited quite a bit, can I get someone who is free to take a look?

File: cmd/kubelet/app/server.go [hunk 1]
```diff
@@ -457,7 +457,7 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.KubeletDeps) (err error) {
 
 	if kubeDeps.CAdvisorInterface == nil {
 		kubeDeps.CAdvisorInterface, err = cadvisor.New(uint(s.CAdvisorPort), s.ContainerRuntime, s.RootDirectory)
-		if err != nil {
+		if err != nil && goruntime.GOOS != "darwin" {
```
Comment (liggitt) [inline cmd/kubelet/app/server.go]: we don't typically do GOOS checks to ignore errors like this. A better option would be to make a platform-specific stubbed implementation of the relevant interface, and use that when GOOS positively identifies a particular platform

Comment (alejandroEsc) [inline cmd/kubelet/app/server.go]: @liggitt thanks for looking into this! That is exactly what I thought and wanted to hear. Do you think the community would be ok with me making such a platform specific interface? it seems like darwin more and more is become a oversight in development. I dont want to dedicate time to something that the community will not be happy to have.

**Comment by liggitt** [comment]
@alejandroEsc also, this isn't an API change. @kubernetes/sig-node-misc would be a better audience

**Comment by alejandroEsc** [comment]
thank you for the awesome suggestion @liggitt! 

Would someone from  @kubernetes/sig-node-misc be kind enough to start a serious conversation about this topic in more detail?

**Comment by euank** [comment]
This will continue to have a chance at breaking unless we test it.
Unfortunately, getting OS X test machines isn't nearly as easy as Linux ones.

cc @kubernetes/sig-testing-misc, thoughts?

**Comment by alejandroEsc** [comment]
@euank I have also noticed that there are gnu specific commands being slowly added to the code base, for example i have seen `cp ... -T` that does not exist in darwin/unix. Are these cases unintentional, or are they on purpose? 

I ask because the current make cross creates a darwin release of k8s, which in the case of cadvisor it is just a broken build. If we are no longer supporting darwin, we probably should just not build for it and make it clear that support is no longer happening. If we do not mean to do this, would it be a problem to add an encompassing issue regarding darwin support?

**Comment by smarterclayton** [comment]
Bash code must be portable to OSX.  We have generally not made the kubelet portable to darwin for lot of reasons.

**Comment by alejandroEsc** [comment]
@smarterclayton
Regarding your first comment: should I then open an issue and try to resolve the particular cases?

Regarding your second: that being said about kubelet, what is the long-term view of kubelet in darwin? k8s in darwin?

**Comment by smarterclayton** [comment]
There is no plan for kubelet on darwin unless someone is willing to commit for it, and it would be a very large chunk of work.  While it would be easier than the windows work, I haven't heard anyone express a desire for it.

For the first issue, please spawn an issue and force people to choose portable bash.

**Comment by alejandroEsc** [comment]
@euank @smarterclayton made changes to at least have a cadvisor interface for darwin. kubelet now runs.

**Comment by euank** [comment]
@alejandroEsc I'd prefer that if `./hack/local-up` were called on darwin it errored out that it's an unsupported platform to fix this, or that darwin becomes a supported platform.

Having `./hack/local-up` succeed, but result in a totally unsupported binary running, seems worse than either of the alternatives.

**Comment by alejandroEsc** [comment]
@euank so a couple of things:
1. I dont think we 'do not support' osx, just as windows? At least that is what i got out of my conversation with @smarterclayton. Only that we need resources to make it fully supported. 
2. The binary succeeds and runs, just that cdavisor keeps warning/complianing.

Could we compromise here and i leave a message warning about not being supported, and letting at least kubelet run as opposed to just crash with not comment?

**Comment by alejandroEsc** [comment]
@vishh please take a look, k8s recommended to contact you as an owner.

**Comment by alejandroEsc** [comment]
I had a great discussion with @euank regarding this PR over slack. The conclusion of which was that 
- Since there are no plans, currently, to support kubelet on darwin we should probably block and make it explicitly known that kubelet running on darwin is not currently supported and so it will not be brought up.
- The appropriate change here is then to create a check for darwin in the script, produce an error message and not attempt to bring up kubelet.

@smarterclayton what are your thoughts on this?
@euank could you please ping the person you indicated would best offer a helpful opinion?

**Comment by euank** [comment]
cc @mikedanese

**Comment by smarterclayton** [comment]
I'm ok with that.

**Comment by vishh** [comment]
Kubernetes isn't supporting darwin nodes yet. Kubelet isn't just the only
node component in k8s. Why are we even trying to create darwin based
clusters?
If its for testing / running e2es, taking a look at #38975

**Comment by smarterclayton** [comment]
We aren't, that was the last comment.

Here's what I would describe:

1. Kubernetes must be buildable on darwin (make build works)
2. Unit tests must pass on darwin (make test works)
3. A dockerized build can be started from darwin (make release build works)
4. A user can start kube-apiserver, kube-controller-manager, and other
components on darwin.  Kubelet is a special case - Kubelet should compile,
but exit with an error, if it cannot run on that platform.  Hollow-Kubelet
*should* start on darwin.

I'll make the change to put these into the dev requirements if @thockin
doesn't decide to make mac users jump off a cliff.

**Comment by alejandroEsc** [comment]
@vishh @smarterclayton @euank made changes we agreed to, ready for review. Thank you!

**Comment by alejandroEsc** [comment]
@k8s-bot gci gke e2e test this

**Pull request merged**